code/components/gta-net-rdr3/MumbleVoice.cpp

```diff
@@ -0,0 +1,694 @@
+    cameraPos[0] = gcc.x;
+    cameraPos[1] = gcc.y;
+    cameraPos[2] = gcc.z;
+
+    //////TO DO  ////////////
+
+    cameraFront[0] = 0.0f;
+    cameraFront[1] = 0.0f;
+    cameraFront[2] = 0.0f;
+
+    cameraTop[0] = 0.0f;
+    cameraTop[1] = 0.0f;
+    cameraTop[2] = 0.0f;
+
+    ////////////////////////
+
+    g_mumbleClient->SetListenerMatrix(actorPos, cameraFront, cameraTop);
+    g_mumbleClient->SetActorPosition(actorPos);
+
+
+    auto likelihoodValue = hook::get_address<uint8_t>(g_preference + 4)  -188;
+
+
+    if (likelihoodValue >= 0 && likelihoodValue < 3)
+    {
+        g_mumbleClient->SetActivationLikelihood(MumbleVoiceLikelihood::VeryLowLikelihood);
+    }
+    else if (likelihoodValue >= 3 && likelihoodValue < 6)
+    {
+        g_mumbleClient->SetActivationLikelihood(MumbleVoiceLikelihood::LowLikelihood);
+    }
+    else if (likelihoodValue >= 6 && likelihoodValue < 9)
+    {
+        g_mumbleClient->SetActivationLikelihood(MumbleVoiceLikelihood::ModerateLikelihood);
+    }
+    else
+    {
+        g_mumbleClient->SetActivationLikelihood(MumbleVoiceLikelihood::HighLikelihood);
+    }
+
+
+    // handle PTT
+    auto isControlPressed = fx::ScriptEngine::GetNativeHandler(0xF3A21BCD95725A4A);
+    fx::ScriptContextBuffer cxt;
+
+    cxt.Push(0);
+    cxt.Push(0x4BC9DABB); // INPUT_PUSH_TO_TALK
+
+    (*isControlPressed)(cxt);
+
+    g_mumbleClient->SetPTTButtonState(cxt.GetResult<bool>());
+
+}
+
+static HookFunction initFunction([]()
+{
+    NetLibrary::OnNetLibraryCreate.Connect([](NetLibrary* netLibrary)
+    {
+        g_netLibrary = netLibrary;
+    });
+    g_mumble.nextConnectDelay = 4 * 1000;
+
+    auto mc = CreateMumbleClient();
+    mc->AddRef();
+    g_mumbleClient = mc.GetRef();
+
+    g_mumbleClient->Initialize();
+
+    g_mumbleClient->SetPositionHook(PositionHook);
+
+    OnMainGameFrame.Connect([=]()
+    {
+        Mumble_RunFrame();
+    });
+
+    OnKillNetworkDone.Connect([]()
+    {
+        g_mumbleClient->SetAudioDistance(0.0f);
+
+        Mumble_Disconnect();
+        o_talkers.reset();
+    });
+});
+
+
+static boost::optional<fx::TNativeHandler> getPlayerName;
+static boost::optional<fx::TNativeHandler> getServerId;
+
+static std::shared_ptr<lab::AudioContext> getAudioContext(int playerId)
+{
+    if (!g_mumble.connected)
+    {
+        return {};
+    }
+
+    std::string name = fmt::sprintf("[%d] %s",
+       FxNativeInvoke::Invoke<int>(getServerId, playerId),
+        FxNativeInvoke::Invoke<const char*>(getPlayerName, playerId));
+
+    return g_mumbleClient->GetAudioContext(name);
+}
+
+static bool(*g_origGetPlayerHasHeadset)(void*, void*);
+
+static bool _getPlayerHasHeadset(void* mgr, void* plr)
+{
+	if (g_mumble.connected)
+	{
+		return 25;
+	}
+
+	return g_origGetPlayerHasHeadset(mgr, plr);
+}
+
+static HookFunction hookFunction([]()
+{
+
+
+    rage::scrEngine::OnScriptInit.Connect([]()
+    {
+
+        g_preference = hook::get_pattern<char>("D0 3E 7B 45 01 00 00 00 E0 4D 27 45", 0x2B4);

```

Comment:
Are you sure this pattern is viable enough? Can't find it on 1207.58 and 1232.62. It seems you hardcoded vtable's address to your pattern, which is not good.

Reply:
I dont have any issues with this pattern i test this code with 4 friends and works fine. But if you want how i can improve this ?

Reply:
That pattern may work fine with the current version of the game which is used by RedM (1207.80 iirc). But it will be broken for newer game patches (which is gonna be ported to RedM I hope). Patterns are used to avoid such breakage between different versions of the game. And your pattern is literally just hardcoded address (0x1457B3ED0?) which will be different for other game versions. In short, you have to find a more viable pattern for that address. Feel free to ping me in CFX's discord and I will try to help you.

Reply:
Memory pattern changed this should be better.

Reply:
The script now uses a pointer to get the address of the voice settings structure

Reply:
Don't you need to mask the memory pattern? Would you need to try different versions of the game and see which bytes stay static and which changes between versions?

Reply:
Memory pattern is valid on both version but 
Disquse say this memory region isn't safe for this